src/pkgcheck/checks/eclass.py

```diff
@@ -498,3 +500,85 @@ def feed(self, eclass):
         )
         if vars_missing_docs:
             yield EclassDocMissingVar(sorted(vars_missing_docs), eclass=eclass)
+
+
+class GoMissingDeps(results.VersionResult, results.Warning):
+    """Package sets ``GO_OPTIONAL`` but does not depend on ``dev-lang/go``."""
+
+    desc = "sets GO_OPTIONAL but does not depend on dev-lang/go"
+
+
+class RubyMissingDeps(results.VersionResult, results.Warning):
+    """Package sets ``RUBY_OPTIONAL`` but does not depend on ``dev-lang/ruby``
+    or ``virtual/rubygems``."""
+
+    desc = "sets RUBY_OPTIONAL but does not depend on dev-lang/ruby or virtual/rubygems"
+
+
+class RustMissingDeps(results.VersionResult, results.Warning):
+    """Package sets ``CARGO_OPTIONAL`` but does not depend on ``virtual/rust``."""
+
+    desc = "sets CARGO_OPTIONAL but does not depend on virtual/rust"
+
+
+class TmpfilesMissingDeps(results.VersionResult, results.Warning):
+    """Package sets ``TMPFILES_OPTIONAL`` but does not depend on ``virtual/tmpfiles``."""
+
+    desc = "sets TMPFILES_OPTIONAL but does not depend on virtual/tmpfiles"
+
+
+class EclassManualDepsCheck(Check):
+    """Check for missing deps when inheriting eclasses in special mode."""
+
+    _source = sources.EbuildParseRepoSource
+    known_results = frozenset(
+        {
+            GoMissingDeps,
+            RustMissingDeps,
+            RubyMissingDeps,
+            TmpfilesMissingDeps,
+        }
+    )
+
+    dependencies = (
+        # eclass, variable, one of deps, class
+        ("cargo", "CARGO_OPTIONAL", {"virtual/rust"}, RustMissingDeps),
+        ("go-module", "GO_OPTIONAL", {"dev-lang/go"}, GoMissingDeps),
+        ("ruby-ng", "RUBY_OPTIONAL", {"dev-lang/ruby", "virtual/rubygems"}, RubyMissingDeps),
```

Comment:
I think this might be wrong for Ruby.

htmltidy has:
```
ruby_add_bdepend "test? ( dev-ruby/thor dev-ruby/tty-editor )"
```

From metadata cache:
```
BDEPEND=ruby_targets_ruby30? ( test? ( dev-ruby/thor[ruby_targets_ruby30(-)] dev-ruby/tty-editor[ruby_targets_ruby30(-)] ) ) ruby_targets_ruby31? ( test? ( dev-ruby/thor[ruby_targets_ruby31(-)] dev-ruby/tty-editor[ruby_targets_ruby31(-)] ) ) ruby_targets_ruby32? ( test? ( dev-ruby/thor[ruby_targets_ruby32(-)] dev-ruby/tty-editor[ruby_targets_ruby32(-)] ) ) >=dev-util/ninja-1.8.2 >=dev-util/cmake-3.20.5
```

Not sure if we should/could check deps-of-deps, or if we should just whitelist dev-ruby/* as well...

Reply:
Fixed, added a way to specify a category `dev-ruby` and it being whitelisted.